### PR TITLE
Fixed #28280 -- Prevented numberformat.format() from formatting large/tiny floats in scientific notation.

### PR DESCRIPTION
**Pull request by wil**
This fix takes into account float values that gets represented in scientific notation.

Please see https://code.djangoproject.com/ticket/28280

### CODE REVIEW

File: tests/utils_tests/test_numberformat.py [hunk 1]
```diff
@@ -57,6 +57,15 @@ def test_decimal_numbers(self):
         self.assertEqual(nformat(Decimal('1234'), '.', grouping=2, thousand_sep=',', force_grouping=True), '12,34')
         self.assertEqual(nformat(Decimal('-1234.33'), '.', decimal_pos=1), '-1234.3')
         self.assertEqual(nformat(Decimal('0.00000001'), '.', decimal_pos=8), '0.00000001')
+        self.assertEqual(nformat(Decimal('.000000000000000000999999'), '.', decimal_pos=2), '0.00')
```
Comment (timgraham) [inline tests/utils_tests/test_numberformat.py]: If you want to add additional tests that already work without this patch, it's okay, but please do it a separate commit so that it's clear which cases this fix is changing.

**Comment by wil** [comment]
I had missed the test case:

```
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python35\tests\i18n\tests.py", line 283, in test_locale_independent
    self.assertEqual('-66666.6', nformat(-66666.666, decimal_sep='.', decimal_pos=1))
AssertionError: '-66666.6' != '-66666.7'
- -66666.6
?        ^
+ -66666.7
?        ^
```

which begs the question of: what's the correct behaviour? Should `format()` perform rounding?

**Comment by timgraham** [comment]
My instinct would be no, it should not do any rounding -- so the expected output for that case would be the original number.

**Comment by wil** [comment]
Thanks Tim. That makes sense. The current implementation does no rounding. I've updated the commit.

**Comment by wil** [comment]
Hi @timgraham would you mind reviewing this again when you get the chance? Thanks.

File: tests/utils_tests/test_numberformat.py [hunk 2]
```diff
@@ -57,6 +61,14 @@ def test_decimal_numbers(self):
         self.assertEqual(nformat(Decimal('1234'), '.', grouping=2, thousand_sep=',', force_grouping=True), '12,34')
         self.assertEqual(nformat(Decimal('-1234.33'), '.', decimal_pos=1), '-1234.3')
         self.assertEqual(nformat(Decimal('0.00000001'), '.', decimal_pos=8), '0.00000001')
+        self.assertEqual(nformat(Decimal('9e-19'), '.', decimal_pos=2), '0.00')
+        self.assertEqual(nformat(Decimal('.00000000000099'), '.', decimal_pos=0), '0')
+        self.assertEqual(nformat(Decimal('3.'), '.'), '3')
+        self.assertEqual(nformat(Decimal('3.0'), '.'), '3.0')
+        self.assertEqual(nformat(Decimal('1e16'), '.', thousand_sep=',', grouping=3, force_grouping=True),
```
Comment (timgraham) [inline tests/utils_tests/test_numberformat.py]: Using hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style). e.g.
```python
self.assertEqual(
    nformat(Decimal('1e16'), '.', thousand_sep=',', grouping=3, force_grouping=True),
    '10,000,000,000,000,000'
)
```

File: django/utils/numberformat.py [hunk 3]
```diff
@@ -42,6 +47,12 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
         int_part, dec_part = str_number, ''
     if decimal_pos is not None:
         dec_part = dec_part + ('0' * (decimal_pos - len(dec_part)))
+    else:
+        if isinstance(number, (float, int)):
+            # {:f} formatted floats get 9 decimal pos by default, strip trailing zeros that add no extra information
```
Comment (timgraham) [inline django/utils/numberformat.py]: Please wrap comments at 79 characters.

File: tests/utils_tests/test_numberformat.py [hunk 4]
```diff
@@ -49,6 +49,10 @@ def test_large_number(self):
         self.assertEqual(nformat(-1 - int_max, '.'), most_max.format('-', '9'))
         self.assertEqual(nformat(-2 * int_max, '.'), most_max2.format('-'))
 
+    def test_float_numbers(self):
+        # The following differs from the Decimal('3.') case because the latter behaves more like an integer
```
Comment (timgraham) [inline tests/utils_tests/test_numberformat.py]: Can you elaborate on "because the latter behaves more like an integer"? I'm not sure exactly what that means.

Comment (wil) [inline tests/utils_tests/test_numberformat.py]: I should have explained more clearly.

I wanted to document the non-obvious difference between passing `numberformat.format()` a float `3.` and a `Decimal(3.)`. In the decimal case, it seems that the "fractional" part is zero, and the result is identical to formatting an int.

```python
>>> nformat(3., '.')
'3.0'
>>> nformat(Decimal('3.'), '.')  # instantiate Decimal with a string
'3'
>>> nformat(Decimal(3.), '.')  # instantiate Decimal with a float
'3'
>>> nformat(3, '.')
'3'
```

File: django/utils/numberformat.py (hunk 3 above)
Comment (wil) [inline django/utils/numberformat.py]: Will do, and I'll fix the hanging indent as well.

**Comment by wil** [comment]
HI @timgraham I've reworked the commits based on your comments.

File: django/utils/numberformat.py [hunk 5]
```diff
@@ -28,6 +28,11 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
     sign = ''
     if isinstance(number, Decimal):
         str_number = '{:f}'.format(number)
+    elif isinstance(number, float):
+        # if decimal_pos is not given, render it with sufficient precision
+        precision = decimal_pos or 8
+        fmt = '{{:.{}f}}'.format(precision + 1)  # gives '{:.1f}' if precision=1
```
Comment (timgraham) [inline django/utils/numberformat.py]: seems like it will be 2f if precision=1 given the +1 after it, right?

File: django/utils/numberformat.py [hunk 6]
```diff
@@ -28,6 +28,11 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
     sign = ''
     if isinstance(number, Decimal):
         str_number = '{:f}'.format(number)
+    elif isinstance(number, float):
+        # if decimal_pos is not given, render it with sufficient precision
+        precision = decimal_pos or 8
```
Comment (timgraham) [inline django/utils/numberformat.py]: What's magic about the value of 8? Seems like it can be changed to other nearby integer values without any tests failing.

File: tests/utils_tests/test_numberformat.py [hunk 7]
```diff
@@ -49,6 +49,19 @@ def test_large_number(self):
         self.assertEqual(nformat(-1 - int_max, '.'), most_max.format('-', '9'))
         self.assertEqual(nformat(-2 * int_max, '.'), most_max2.format('-'))
 
+    def test_float_numbers(self):
+        self.assertEqual(nformat(0.00000001, '.', decimal_pos=8), '0.00000001')
+        self.assertEqual(nformat(9e-19, '.', decimal_pos=2), '0.00')
+        self.assertEqual(nformat(.00000000000099, '.', decimal_pos=0), '0')
+        self.assertEqual(nformat(1e16, '.', thousand_sep=',', grouping=3, force_grouping=True),
+                         '10,000,000,000,000,000.0')
```
Comment (timgraham) [inline tests/utils_tests/test_numberformat.py]: Use hanging indent as mentioned before.

File: django/utils/numberformat.py [hunk 8]
```diff
@@ -42,6 +47,13 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
         int_part, dec_part = str_number, ''
     if decimal_pos is not None:
         dec_part = dec_part + ('0' * (decimal_pos - len(dec_part)))
+    else:
+        if isinstance(number, (float, int)):
```
Comment (timgraham) [inline django/utils/numberformat.py]: Doesn't see a test failure if `int` is removed here -- is a test missing?

**Comment by timgraham** [comment]
@wil, do you have time to finish this up?

**Comment by wil** [comment]
@timgraham thanks for reminding me. I'll work on it in the next few days.

**Comment by rixx** [comment]
@wil Are you still interested in pursuing this PR?

**Comment by felixxm** [comment]
Closing due to inactivity.
